stdlib/moreLabels.mli

```diff
@@ -1321,12 +1321,97 @@ module Set : sig
       val of_seq : elt Seq.t -> t
       (** Build a set from the given bindings
           @since 4.07 *)
+
+      (** Enumerations of the elements of a set.
+
+          An enumeration is an immutable data structure: its implementation
+          involves no side effects. It represents the sequence of the elements
+          of a set, in increasing order. In addition to enumerating the elements
+          of the sequence (one by one), it supports skipping ahead in the
+          sequence to the point where a certain threshold is reached or
+          exceeded.
+          @since 5.3 *)
+      module Enum : sig
+
+        type set = t
+        (** The type [set] is a synonym for the type of sets.
+            @since 5.3 *)
+
+        type enum
+        (** The type of enumerations. An enumeration represents an increasing
+            sequence of elements of type [elt]. The sequence is increasing
+            according to [Ord.compare]: thus, if [x] appears earlier than [y] in
+            the sequence, then [Ord.compare x y < 0] holds.
+            @since 5.3 *)
+
+        type t = enum
+        (** A synonym for the type [enum].
+            @since 5.3 *)
+
+        val empty : enum
+        (** [empty] is the empty enumeration. It contains zero elements.
+            @since 5.3 *)
+
+        val is_empty : enum -> bool
+        (** [is_empty e] tests whether the enumeration [e] is empty.
+            @since 5.3 *)
+
+        val enum : set -> enum
+        (** [enum s] returns an enumeration of the set [s]. This enumeration
+            contains all of the elements of the set [s], in increasing order.
+            @since 5.3 *)
+
+        val enum_from : elt -> set -> enum
+        (** [enum_from x s] returns an enumeration of the subset of [s]
+            formed of just the elements that are no less than [x].
+            It is equivalent to [from x (enum s)].
+            @since 5.3 *)
+
+        val head : enum -> elt
+        (** [head e] returns the first element of the enumeration [e].
+            The enumeration [e] must be nonempty.
+            @since 5.3 *)
```

Comment:
I would rather call this `get_head` rather than `head` to emphasize that it is an operation that can fail. The kind of exception raised should be documented (Invalid_argument; it could have been Not_found). Same with `tail` of course.

No strong opinion on how the corresponding `head_opt` function should be named; maybe keeping `head_opt` is fine, maybe staying consistent with `get_head_opt` is wiser.

Reply:
It seems there is still a lot of confusion about the naming conventions of getters here. Of course the stdlib is not entirely consistent. But I think it would be good if we gradually move to this for getting a property `prop` from a structure:

1. If people really want to add a raising variant with an ad-hoc exception then this should be:
```ocaml
val prop : t -> a (* raises some ad-hoc exception if unavailable like Not_found *)
val get_prop : t -> a  (* raises Invalid_argument if unavailable *)
val prop_opt : t -> a option
```

2. If people are ok with not having an ad-hoc raising variant. Which I think can be fine in some new interfaces: 
```ocaml
val prop : t -> a option
val get_prop : t -> a (* raises Invalid_argument if unavailable. *)
```

3. If people are ok with not having and ad-hoc raising variant and trust functional programming: 
    ```ocaml
    val prop : t -> a option
    ```
    and simply use `M.prop v |> Option.get` for when they know by an invariant that they expect a some.

Another way of saying is that if you have an `_opt` in the name you don't have a `get_` in the name. If you have a `get_` in the name it means you can have an `Invalid_argument` thrown in your face (I think this is mostly consistent in the stdlib).

If you ask me in this case I would, like I'm advocating for https://github.com/ocaml/ocaml/pull/12871 just go for 3. If people disagree I still think that either 1. or 2. should be religiously followed.

Reply:
Thanks! I'm indeed myself confused about these considerations. Maybe we should write this down in stdlib/HACKING.adoc or somewhere, if there is consensus.